File: AJS_MGLMM_Overdispersed_Count/Code_as_It_Is/04compoisson_amostra_R.cpp
// COM-Poisson full
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  using namespace density;
  DATA_VECTOR(Y1);
  DATA_VECTOR(Y2);
  DATA_MATRIX(X);
  PARAMETER_VECTOR(beta1);
  PARAMETER_VECTOR(beta2);
  PARAMETER_MATRIX(U);
  PARAMETER(rho);
  PARAMETER_VECTOR(sigma);
  PARAMETER_VECTOR(nu);
  int n = Y1.size();
  int u = U.cols();
  Type nll = 0;
  // Response Variables
  vector<Type> mu1 = X*beta1 + U.col(0).array();
  for(int i = 0; i < n; i++)
    nll -= dcompois2(Y1(i), exp(mu1(i)), exp(nu(0)), true);
  vector<Type> mu2 = X*beta2 + U.col(1).array();
  for(int i = 0; i < n; i++)
    nll -= dcompois2(Y2(i), exp(mu2(i)), exp(nu(1)), true);
  // Reparametrization of input parameters
  sigma(0) = pow(exp(sigma(0)),2);
  sigma(1) = pow(exp(sigma(1)),2);
  rho = -1 + 2*(exp(rho)/(exp(rho) + 1)); // // transf.rho <- function(rho){-1+2*(exp(rho)/(exp(rho)+1))}
  // Random effect
  Type cov_rho = rho*sqrt(sigma(0))*sqrt(sigma(1));
  matrix<Type> Sigma(u,u);
  Sigma.row(0) << sigma(0), cov_rho;
  Sigma.row(1) << cov_rho, sigma(1);
  MVNORM_t<Type> neg_log_density(Sigma); // Create object from covariance matrix Sigma; (neg_log_density could be any name)
  for (int i=0; i<n; i++){
    nll += neg_log_density(U.row(i)); // Process likelihood
    SIMULATE {
      U.row(i) = neg_log_density.simulate();
    }
  }
  SIMULATE {
  vector<Type> mu1 = X*beta1 + U.col(0).array();
  Y1 = rcompois2(exp(mu1), exp(nu(0)));  // Simulate response
  REPORT(Y1);          // Report the simulation
  
  vector<Type> mu2 = X*beta2 + U.col(1).array();
  Y2 = rcompois2(exp(mu2), exp(nu(1)));  // Simulate response
  REPORT(Y2);          // Report the simulation
  }

  ADREPORT(exp(nu(0)));
  ADREPORT(exp(nu(1)));
  ADREPORT(exp(sigma(0))); // Standard Deviation
  ADREPORT(exp(sigma(1))); // Standard Deviation
  ADREPORT(-1 + 2*(exp(rho)/(exp(rho) + 1)));

  // for(int i = 0; i < n; i++){
  //   nll += VECSCALE(UNSTRUCTURED_CORR(rho), sigma)(U.row(i)); // Likelihood
  // }
  // 
  // matrix<Type> Cor(2,2);
  // Cor = UNSTRUCTURED_CORR(rho).cov();
  // ADREPORT(Cor);
  
  
  return nll;
}

File: AJS_MGLMM_Overdispersed_Count/Code_as_It_Is/04compoisson_amostra_R.R
## Compoisson distribution -----------------------------------------------------
## Authors: Multivariate Statistical Modelling Group ---------------------------
## Date: June, 18 2020 ---------------------------------------------------------
devtools::install_github("thomas-fung/mpcmp")
rm(list=ls())
require(MASS)
require(mvtnorm)
require(TMB)
require(compoisson)
## Data simulation -------------------------------------------------------------
set.seed(1234)
beta1 <- c(.2, 0.8)
beta2 <- c(0.5, -1)
n <- 500
x1 <- rnorm(n)
X <- model.matrix(~x1)
## Random effects covariance matrix
s1 <- log(0.5)
s2 <- log(0.3)
true_rho <- 0
rho <- 0.5*log((1 + true_rho)/ (1-true_rho))
Sigma <- matrix(NA, 2, 2)
Sigma[1,1] <- exp(s1)
Sigma[2,2] <- exp(s2)
Sigma[1,2] <- rho*sqrt(exp(s1))*sqrt(exp(s2))
Sigma[2,1] <- rho*sqrt(exp(s1))*sqrt(exp(s2))
set.seed(1234)
U <- rmvnorm(n, mean = c(0,0), sigma = Sigma)
cor(U)
mu1 <- exp(X%*%beta1 + U[,1])
mu2 <- exp(X%*%beta2 + U[,2])
## Response variables
nu1 <- .7
nu2 <- .7
l_nu1 <- log(nu1)
l_nu2 <- log(nu2)
set.seed(1234)
(Y1 <- mpcmp::rcomp(n, mu = mu1, nu = nu1))
set.seed(1234)
(Y2 <- mpcmp::rcomp(n, mu = mu2, nu = nu2))
# data_export_R <- list(Y1 = Y1,
#                       Y2 = Y2,
#                       X = X)
# saveRDS(data_export_R, file = "com_poisson_data_R.rds")
## Data set
data <- list(Y1 = Y1,
             Y2 = Y2,
             X = X)
## Parameters
parameters <- list(beta1 = beta1,
                   beta2 = beta2,
                   U = matrix(0, ncol = 2, nrow = n),
                   rho = true_rho,
                   sigma = c(sqrt(exp(s1)), sqrt(exp(s2))),
                   nu = c(l_nu1, l_nu2))

## Compilation
setwd("~/Google Drive/Mestrado/dissertacao/TMB/MultivariateModels")
model <- "04_compoisson_bi_full"
compile(paste0(model, ".cpp"), flags = "-O0 -ggdb")
dyn.load(dynlib(model))
openmp(n=2)
obj <- MakeADFun(data, parameters, DLL=model, random = "U")
system.time(fit_NLB <- nlminb(start = obj$par, objective = obj$fn, gradient = obj$gr))
## Obtaining parameters in a logical way
ss <- sdreport(obj)
rep_fixed <- summary(ss, "fixed")
rep_report <- summary(ss, "report")
source("/home/guilherme/Google Drive/Mestrado/dissertacao/TMB/MultivariateModels/Servidor/unify_packages.R")
round(TMB_summary(rep_fixed, rep_report, 2), 4)
